pythonforandroid/recipe.py

```diff
@@ -479,18 +498,42 @@ def apply_patches(self, arch, build_dir=None):
 
     def should_build(self, arch):
         '''Should perform any necessary test and return True only if it needs
-        building again.
+        building again. Per default we implement a library test, in case that
+        we detect so.
 
         '''
+        if self.built_libraries:
+            return not all(
+                exists(lib) for lib in self.get_libraries(arch.arch)
+            )
         return True
 
+    def do_build_libs(self, arch):
+        '''The system library recipes, defined by attribute
+        :attr:`~pythonforandroid.recipe.Recipe.built_libraries`, should
+        overwrite this class with the proper commands to compile the libraries
+        '''
+        if self.built_libraries:
+            raise NotImplementedError("Not implemented in Recipe")
+
+    def do_install_libs(self, arch):
+        shared_libs = [
+            lib for lib in self.get_libraries(arch) if lib.endswith(".so")
+        ]
+        self.install_libs(arch, *shared_libs)
+
     def build_arch(self, arch):
         '''Run any build tasks for the Recipe. By default, this checks if
         any build_archname methods exist for the archname of the current
         architecture, and runs them if so.'''
         build = "build_{}".format(arch.arch)
         if hasattr(self, build):
             getattr(self, build)()
+        # in case that we detect a library recipe, we split the build in two
+        # steps, so we can control the copy and check of the built libraries
+        if self.built_libraries:
+            self.do_build_libs(arch)
+            self.do_install_libs(arch)
```

Comment:
There's nothing really wrong with this, but I mildly dislike replacing `build_arch` with `do_build_libs` in a subset of recipes.

I think it could be better to add an `install_libs` method that called after build_arch in `build.build_recipes`, and which does the job currently covered by `do_install_libs`. That way, all the recipes could continue to use `build_arch` the same way they currently do, which is nice for consistency.